server/block/sugar_cane.go

```diff
@@ -0,0 +1,111 @@
+package block
+
+import (
+	"github.com/df-mc/dragonfly/server/block/cube"
+	"github.com/df-mc/dragonfly/server/item"
+	"github.com/df-mc/dragonfly/server/world"
+	"github.com/df-mc/dragonfly/server/world/particle"
+	"github.com/go-gl/mathgl/mgl64"
+	"math/rand"
+)
+
+// SugarCane is a plant block that generates naturally near water.
+type SugarCane struct {
+	empty
+	transparent
+
+	// Age is the growth state of suger cane. Values range from 0 to 15.
+	Age int
+}
+
+// UseOnBlock handles making sure a neighbouring blocks contains water.
+func (c SugarCane) UseOnBlock(pos cube.Pos, face cube.Face, _ mgl64.Vec3, w *world.World, user item.User, ctx *item.UseContext) (used bool) {
+	pos, _, used = firstReplaceable(w, pos, face, c)
+	if !used {
+		return false
+	}
+	if !c.CanGrowHere(pos, w) {
+		return false
+	}
+
+	place(w, pos, c, user, ctx)
+	return placed(ctx)
+}
+
+// NeighbourUpdateTick ...
+func (c SugarCane) NeighbourUpdateTick(pos, _ cube.Pos, w *world.World) {
+	if !c.CanGrowHere(pos, w) {
+		w.SetBlock(pos, nil, nil)
+		w.AddParticle(pos.Vec3Centre(), particle.BlockBreak{Block: c})
+	}
+}
+
+// RandomTick ...
+func (c SugarCane) RandomTick(pos cube.Pos, w *world.World, r *rand.Rand) {
+	if c.Age < 15 {
+		c.Age++
+	} else if c.Age == 15 {
+		c.Age = 0
+		if c.CanGrowHere(pos.Side(cube.FaceDown), w) {
+			for y := 1; y < 3; y++ {
+				if _, ok := w.Block(pos.Add(cube.Pos{0, y})).(Air); ok {
+					w.SetBlock(pos.Add(cube.Pos{0, y}), SugarCane{Age: 0}, nil)
+					break
+				} else if _, ok := w.Block(pos.Add(cube.Pos{0, y})).(SugarCane); !ok {
+					break
+				}
+			}
```

Comment:
Will this not make it grow way too fast? In vanilla only the top sugar cane ever ages, so this logic doesn't exactly match

Reply:
Taken from the Cactus code. it technically only grows when the base is randomly ticked.